### ISSUE TITLE
Attention Mechanisms and Memory Networks

### ISSUE THREAD

**Issue by mac40** [Mmiglio/SpeechRecognition#3]
* [x] Study
* [x] Implement RNN
    * [x] LSTM
    * [x] GRU
    * [ ] Autoencoders
* [x] Test
    * [x] activate/deactivate Attention
    * [x] dropout
    * [x] double vs single RNN layer
    * [x] visualize attention scores

Useful references:
* [A Beginner's Guide to Attention Mechanisms and Memory Networks](https://skymind.ai/wiki/attention-mechanism-memory-network)
* [Intuitive Understanding of Attention Mechanism in Deep Learning](https://towardsdatascience.com/intuitive-understanding-of-attention-mechanism-in-deep-learning-6c9482aecf4f)
* [Tutorial on Attention-based Models (Part 1)](https://krntneja.github.io/blog/2018/06/02/attention-based-models-1)
* [Show, Attend and Tell: Neural Image Caption Generation with Visual Attention](https://arxiv.org/pdf/1502.03044v3.pdf)

**Comment by mac40** [comment]
# Useful documentation for attention implementation:

* [Attention in Deep Networks with Keras](https://towardsdatascience.com/light-on-math-ml-attention-with-keras-dc8dbc1fad39)
* [A neural attention model for speech command recognition](https://www.researchgate.net/publication/327260186_A_neural_attention_model_for_speech_command_recognition)
* [Keras LSTM tutorial](https://adventuresinmachinelearning.com/keras-lstm-tutorial/)
* [Understanding LSTM](https://towardsdatascience.com/understanding-lstm-and-its-quick-implementation-in-keras-for-sentiment-analysis-af410fd85b47)
* [Deep Learning #4: Why You Need to Start Using Embedding Layers](https://towardsdatascience.com/deep-learning-4-embedding-layers-f9a02d55ac12)
* [Speech Recognition with Neural Networks](http://dkopczyk.quantee.co.uk/speech-nn/)

# documentation for attention visualization
* [How to Visualize Your Recurrent Neural Network with Attention in Keras](https://medium.com/datalogue/attention-in-keras-1892773a4f22)
* [keras-attention repo](https://github.com/datalogue/keras-attention)
* [How to visualize attention weights?](https://stackoverflow.com/questions/53867351/how-to-visualize-attention-weights)

# documentation for RNN autoencoders
* [Step-by-step understanding LSTM Autoencoder layers](https://towardsdatascience.com/step-by-step-understanding-lstm-autoencoder-layers-ffab055b6352)
* [How to Use the TimeDistributed Layer in Keras](https://machinelearningmastery.com/timedistributed-layer-for-long-short-term-memory-networks-in-python/)

**Comment by mac40** [comment]
Pausing issue before closure unless autoencoders automagically solve themselves
